msmarco-v2-vector/track.py

```diff
@@ -51,5 +136,103 @@ def params(self):
         return result
 
 
+class KnnRecallParamSource:
+    def __init__(self, track, params, **kwargs):
+        if len(track.indices) == 1:
+            default_index = track.indices[0].name
+        else:
+            default_index = "_all"
+
+        self._index_name = params.get("index", default_index)
+        self._cache = params.get("cache", False)
+        self._params = params
+        self.infinite = True
+
+    def partition(self, partition_index, total_partitions):
+        return self
+
+    def params(self):
+        return {
+            "index": self._index_name,
+            "cache": self._params.get("cache", False),
+            "size": self._params.get("k", 10),
+            "num_candidates": self._params.get("num-candidates", 100),
+            "num_rescore": self._params.get("num-rescore", 0),
+        }
+
+
+class KnnRecallRunner:
```

Comment:
This runner compute the relevance metrics on two different "judgment lists":
- ndcg is computed from qrels
- recall is computed from the exact knn as a source of truth

It can be misleading and it took me a while to understand why you can not rely on the qrel to compute the recall. 
My understanding is that you are interested in the differences between exact and approx knn search for the recall.

I would personally split it using two runners at least in a follow up. I would also compute a recall from the judgement list.

Reply:
Discussed offline, we'll address the cleanup in a follow up.